File: py/1296-divide-array-in-sets-of-k-consecutive-numbers.py
from collections import Counter


class Solution:
    def isPossibleDivide(self, nums: list[int], k: int) -> bool:
        counts = Counter(nums)

        for start in sorted(counts):
            start_freq = counts[start]
            if start_freq > 0:
                for i in range(start, start + k):
                    counts[i] -= start_freq
                    if counts[i] < 0:
                        return False

        return True

File: cpp/1296-divide-array-in-sets-of-k-consecutive-numbers.cpp
#include <map>
#include <vector>

class Solution {
    public:
        [[nodiscard]] auto
        isPossibleDivide(const std::vector<int> &nums, const int k) -> bool {
            auto counts = std::map<int, int>{};
            for (const auto card : nums) {
                ++counts[card];
            }

            for (const auto [start, start_freq] : counts) {
                if (start_freq > 0) {
                    for (auto i = start; i < start + k; ++i) {
                        counts[i] -= start_freq;
                        if (counts[i] < 0) {
                            return false;
                        }
                    }
                }
            }

            return true;
        }
};
